### PR TITLE
New API endpoint to show a list of files in an IPFS directory

### PR DESCRIPTION
**Pull request by mbrizic**

### CODE REVIEW

File: nodes/persistence-node/src/services/IpfsGatewayApi.ts [hunk 1]
```diff
@@ -64,11 +72,16 @@ export class IpfsGatewayApi {
 
     app.get('/api/v0/cat', handleError(async (req, res) => {
       const hash = req.query.arg as string;
+      const filename = req.query.filename as string;
```
Comment (nerfZael) [inline nodes/persistence-node/src/services/IpfsGatewayApi.ts:75]: this isn't a required query argument

File: nodes/persistence-node/src/services/IpfsGatewayApi.ts [hunk 2]
```diff
@@ -78,33 +91,51 @@ export class IpfsGatewayApi {
...
+      const resolvedPath = await ipfs.resolve(`/ipfs/${hash}`);
 
       res.json({
         path: resolvedPath
       });
     }));
 
-    app.post('/add', upload.fields([ { name: "files"}, { name: "options", maxCount: 1 } ]), handleError(async (req, res) => {
-      if(!req.files) {
+    app.get('/api/v0/ls/:hash', handleError(async (req, res) => {
```
Comment (nerfZael) [inline nodes/persistence-node/src/services/IpfsGatewayApi.ts:109]: IPFS does have an endpoint like this https://docs.ipfs.io/reference/http/api/#api-v0-ls
But it does something different
Rename this endpoint to '/ipfs/:hash' to mimick ipfs.io/ipfs/:hash

File: nodes/persistence-node/src/utils/toArray.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export async function toArray<T>(iterable: AsyncIterable<T>): Promise<T[]> {
```
Comment (nerfZael) [inline nodes/persistence-node/src/utils/toArray.ts:1]: This isn't very descriptive, maybe 'asyncIterableToArray' would be more appropriate?

**Review by nerfZael — CHANGES_REQUESTED** [review]

File: nodes/persistence-node/src/services/IpfsGatewayApi.ts (hunk 1 above)
Comment (mbrizic) [reply nodes/persistence-node/src/services/IpfsGatewayApi.ts:75]: Should we make it required? If we omit it, file download just returns the file with file name `cat`.

Comment (mbrizic) [reply nodes/persistence-node/src/services/IpfsGatewayApi.ts]: Fixed.

File: nodes/persistence-node/src/services/IpfsGatewayApi.ts [hunk 4]
```diff
@@ -65,10 +73,14 @@ export class IpfsGatewayApi {
     app.get('/api/v0/cat', handleError(async (req, res) => {
       const hash = req.query.arg as string;
 
-      const stream = ipfs.cat(hash); 
+      if (hash == null) {
```
Comment (nerfZael) [inline nodes/persistence-node/src/services/IpfsGatewayApi.ts:76]: can this be undefined? maybe use !hash ?

**Review by nerfZael — CHANGES_REQUESTED** [review]

File: nodes/persistence-node/src/services/IpfsGatewayApi.ts (hunk 4 above)
Comment (mbrizic) [reply nodes/persistence-node/src/services/IpfsGatewayApi.ts:76]: Changed both. 

- `!hash` should also cover empty strings, which null check didn't do.
- throwing is used so error ends up at global error handler, which might not be needed here, but I wouldn't avoid that approach in general

**Review by nerfZael — CHANGES_REQUESTED** [review]

File: nodes/persistence-node/src/services/IpfsGatewayApi.ts (hunk 4 above)
Comment (nerfZael) [reply nodes/persistence-node/src/services/IpfsGatewayApi.ts:76]: Also, I don't see the `!hash` change. Maybe you didn't push :D

**Review by nerfZael — APPROVED** [review]

**Pull request merged**